com.unity.multiplayer.mlapi/Runtime/Timing/NetworkTimeSystem.cs

```diff
@@ -0,0 +1,111 @@
+    /// </summary>
+    public class NetworkTimeSystem
+    {
+        private double m_Time;
+
+        private double m_CurrentLocalTimeOffset;
+
+        private double m_DesiredLocalTimeOffset;
+
+        private double m_CurrentServerTimeOffset;
+
+        private double m_DesiredServerTimeOffset;
+
+        /// <summary>
+        /// Gets or sets the amount of time in seconds the server should buffer incoming client messages.
+        /// This increases the difference between local and server time so that messages arrive earlier on the server.
+        /// </summary>
+        public double LocalBuffer { get; set; }
+
+        /// <summary>
+        /// Gets or sets the amount of the time in seconds the client should buffer incoming messages from the server. This increases server time.
+        /// A higher value increases latency but makes the game look more smooth in bad networking conditions.
+        /// </summary>
+        public double ServerBuffer { get; set; }
+
+        /// <summary>
+        /// Gets or sets a threshold in seconds used to force a hard catchup of network time.
+        /// </summary>
+        public double HardResetThreshold { get; set; }
+
+        /// <summary>
+        /// Gets or sets the ratio at which the NetworkTimeSystem speeds up or slows down time.
+        /// </summary>
+        public double AdjustmentRatio { get; set; }
+
+        public double LocalTime => m_Time + m_CurrentLocalTimeOffset;
+
+        public double ServerTime => m_Time + m_CurrentServerTimeOffset;
+
+        //TODO This is used as a workaround to pass this back into the sync function will be removed once we get correct value.
+        internal double TimeSystemInternalTime => m_Time;
+
+        public NetworkTimeSystem(double localBuffer, double serverBuffer, double hardResetThreshold, double adjustmentRatio = 0.01d)
+        {
+            LocalBuffer = localBuffer;
+            ServerBuffer = serverBuffer;
+            HardResetThreshold = hardResetThreshold;
+            AdjustmentRatio = adjustmentRatio;
+        }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="NetworkTimeSystem"/> class for a server instance.
+        /// The server will not apply any buffer values which ensures that local time equals server time.
+        /// </summary>
+        /// <returns>The instance.</returns>
+        public static NetworkTimeSystem ServerTimeSystem()
+        {
+            return new NetworkTimeSystem(0, 0, double.MaxValue);
+        }
+
+        /// <summary>
+        /// Advances the time system by a certain amount of time. Should be called once per frame with Time.deltaTime or similar.
+        /// </summary>
+        /// <param name="deltaTime">The amount of time to advance. The delta time which passed since Advance was last called.</param>
+        /// <returns></returns>
+        public bool Advance(double deltaTime)
```

Comment:
where is this being called from atm? (and also `NetworkTickSystem.UpdateTick(localTime, serverTime)`)?
I couldn't see where `NetworkTickSystem` and `NetworkTimeSystem` are integrated into `NetworkManager` or `NetworkUpdateLoop` etc.
How does all this is supposed work currently? Did any changes already went separately from this PR?

Reply:
![image](https://user-images.githubusercontent.com/43687322/125774440-6054a873-06a8-4c87-a939-0d8ca63311b5.png)

In NetworkManager.OnNetworkPreUpdate.

As for how it currently works. The time advances but the network statistic values are not correctly synchronized with the time system yet. So the time system doesn't properly adjust to changes in network conditions yet.

Reply:
The above comment from me is no longer fully update as the network stats are now correctly updated by the time sync message.